benchmarks/spiegelman_et_al_2016_gcubed/spiegelman_material.cc

```diff
@@ -0,0 +1,680 @@
+     * $\rho(T)=\rho_0(1-\beta(T-T_{\text{ref}}))$.
+     *
+     *
+     * @ingroup MaterialModels
+     */
+
+    template <int dim>
+    class DruckerPragerCompositions : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
+    {
+      public:
+        std::vector<double> compute_volume_fractions( const std::vector<double> &compositional_fields) const;
+
+        double compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const;
+
+        double compute_viscosity(const double edot_ii,const double pressure,const int comp, const double prefactor,const bool regularize, const double min_visc, const double max_visc) const;
+
+        virtual void evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+                              MaterialModel::MaterialModelOutputs<dim> &out) const;
+
+        /**
+         * Return whether the model is compressible or not.  Incompressibility
+         * does not necessarily imply that the density is constant; rather, it
+         * may still depend on temperature or pressure. In the current
+         * context, compressibility means whether we should solve the continuity
+         * equation as $\nabla \cdot (\rho \mathbf u)=0$ (compressible Stokes)
+         * or as $\nabla \cdot \mathbf{u}=0$ (incompressible Stokes).
+        *
+        * This material model is incompressible.
+         */
+        virtual bool is_compressible () const;
+
+        virtual double reference_viscosity () const;
+
+        virtual double reference_density () const;
+
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+
+        double reference_T;
+        std::vector<double> reference_T_list;
+
+        /**
+         * Defining a minimum strain rate stabilizes the viscosity calculation,
+         * which involves a division by the strain rate. Units: $1/s$.
+         */
+        bool use_deviator_of_strain_rate;
+        std::vector<double> min_strain_rate;
+        std::vector<double> min_visc;
+        std::vector<double> max_visc;
+        std::vector<double> veff_coefficient;
+        double ref_visc;
+        double reference_compressibility;
+        std::vector<double> ref_visc_list;
+
+        std::vector<double> thermal_diffusivity;
+        std::vector<double> heat_capacity;
+
+
+        std::vector<double> densities;
+        std::vector<double> thermal_expansivities;
+
+        std::vector<double> cohesion;
+        std::vector<double> phi;
+
+        std::vector<double> prefactor;
+        std::vector<double> stress_exponent;
+
+        unsigned int n_fields;
+        // averaging parameters
+        double viscosity_averaging_p;
+
+        bool use_analytical_derivative;
+
+        //optimize variables
+        const double sqrt_3 = std::sqrt(3);
+        const double sqrt_half = std::sqrt(0.5);
```

Comment:
Just use the exact formula in all of the places where you use the variables now. The compiler can optimize these things. (This is also not valid C++98, though it's valid C++11.)